lib/ecto/parameterized_type.ex

```diff
@@ -0,0 +1,83 @@
+defmodule Ecto.ParameterizedType do
+  @moduledoc """
+  Parameterized types are similar to Ecto Custom Types, however they allow a set of options to be
+  specified in the schema which are initialized on compilation and passed to the callback functions
+  as a final argument. Note that options should be specified as a keyword, but it is idiomatic to
+  convert them to maps inside init for easier pattern matching in other callbacks.
+
```

Comment:
We should also a paragraph that explains that, opposite to regular types, custom types are responsible for handling `nil` in both `load` and `dump` callbacks.